lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs

```diff
@@ -72,30 +73,91 @@ pub struct TeleportRdpdrBackend {
     active_device_ids: Vec<u32>,
     /// The client handle for this backend, used to send messages to the RDP server.
     client_handle: ClientHandle,
+    // contexts holds all the active contexts for the server, established using
+    // SCARD_IOCTL_ESTABLISHCONTEXT. Some IOCTLs are context-specific and pass it as argument.
+    //
+    // contexts also holds a cache and connected smartcard handles for each context.
+    contexts: Contexts,
 }
 
 impl TeleportRdpdrBackend {
     pub fn new(smartcard_device_id: u32, client_handle: ClientHandle) -> Self {
         Self {
             active_device_ids: vec![smartcard_device_id],
             client_handle,
+            contexts: Contexts::new(),
         }
     }
 
     fn get_scard_device_id(&self) -> PduResult<u32> {
-        if self.active_device_ids.len() == 0 {
+        if self.active_device_ids.is_empty() {
             return Err(other_err!(
                 "TeleportRdpdrBackend::get_scard_device_id",
                 "no active devices",
             ));
         }
         Ok(self.active_device_ids[0])
     }
+
+    fn handle_access_started(
+        &mut self,
+        req: DeviceControlRequest<ScardIoCtlCode>,
+        _call: ScardAccessStartedEventCall,
+    ) -> PduResult<()> {
+        if req.header.device_id != self.get_scard_device_id()? {
+            return Err(other_err!(
+                "TeleportRdpdrBackend::handle_scard_access_started_event_call",
+                "got ScardAccessStartedEventCall for unknown device_id",
+            ));
+        }
+
+        self.write_rdpdr_dev_ctl_resp(req, Box::new(LongReturn::new(ReturnCode::Success)))
+            .map_err(|_e| {
+                other_err!(
+                    "TeleportRdpdrBackend::handle_access_started",
```

Comment:
Do these errors get surfaced to the user? I think this type of language is okay for logs, but it's not very user friendly.

Reply:
Currently none of these are getting surfaced to the user. Exactly how we should handle Rust errors is still a somewhat open question in my mind, @probakowski has some suggestion in https://github.com/gravitational/teleport/pull/32925 (which I just remembered I need to review).